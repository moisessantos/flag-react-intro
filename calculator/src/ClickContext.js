import React from 'react';

export const ClickContext = React.createContext();

export const useClickContext = () => {
  const context = React.useContext(ClickContext);
  if (!context) {
    throw new Error('useClickContext must be used within the ClickContextProvider');
  }
  return context;
};
