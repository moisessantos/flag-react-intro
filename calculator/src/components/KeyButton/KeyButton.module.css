.operator {
    background-color: #ff6700;
}