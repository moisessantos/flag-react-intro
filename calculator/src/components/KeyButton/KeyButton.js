import React, { memo } from 'react';
import PropTypes from 'prop-types';
import Button from 'react-bootstrap/Button';
import { useClickContext } from '../../ClickContext';
import styles from './KeyButton.module.css';


const KeyButton = ({ text, ...otherProps }) => {
  const handleClick = useClickContext();
  console.warn("render");
  return (
    <Button
      variant="secondary"
      name={text}
      onClick={e => handleClick(e.target.name)}
      {...otherProps}
      className={isNaN(Number(text)) && styles.operator}>
      {text}
    </Button>);
}

KeyButton.propTypes = {
  text: PropTypes.string.isRequired,
};

export default memo(KeyButton, (prevProps, nextProps) => prevProps.text === nextProps.text);
