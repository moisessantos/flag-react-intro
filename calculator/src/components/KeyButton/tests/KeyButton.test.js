import React from 'react';
import { shallow } from 'enzyme';
import KeyButton from '../';
import * as ClickContext from '../../../ClickContext';

jest.mock('../../../ClickContext');

describe('KeyButton tests', () => {
    it('should render as expected', () => {
        const wrapper = shallow(<KeyButton text='5' />)
        expect(wrapper).toMatchSnapshot();
    });

    it('should call handleClick on button click', () => {
        const mock = jest.fn();
        const event = { target: { name: '5' } };
        ClickContext.useClickContext = () => mock;
        const wrapper = shallow(<KeyButton text='5' />);
        wrapper.find('Button').simulate('click', event);
        expect(mock).toBeCalledTimes(1);
        expect(mock).toBeCalledWith(event.target.name);
    });
})