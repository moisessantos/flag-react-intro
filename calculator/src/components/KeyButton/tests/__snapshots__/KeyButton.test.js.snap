// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KeyButton tests should render as expected 1`] = `
<Button
  active={false}
  className={false}
  disabled={false}
  name="5"
  onClick={[Function]}
  type="button"
  variant="secondary"
>
  5
</Button>
`;
